artemis/reporting/base/normal_form.py

```diff
@@ -0,0 +1,92 @@
+import urllib.parse
+from typing import Tuple
+
+from artemis.config import Config
+
+# A normal form of a single vulnerability report. This should be generated in such a way that
+# vulnerabilities that should be deduplicated (e.g. exposed http://example.com/wp-config.php.bak
+# and https://www.example.com/wp-config.php.bak) have the same normal form.
+NormalForm = Tuple[Tuple[str, str], ...]
+
+
+def get_url_normal_form(url: str) -> str:
+    """
+    Normalizes a URL.
+
+    If it's on a standard port we set the port to 0 and set the schema to http_or_https so that
+    http://service.com:80/ and https://service.com:443/ will get the same normal form.
+
+    We also normalize the domain by stripping www.
+    """
+
+    assert "://" in url
+
+    url_parsed = urllib.parse.urlparse(url)
+    url_parsed_dict = url_parsed._asdict()
+    assert url_parsed.hostname
+
+    port = url_parsed.port
+
+    if not port:
+        if url_parsed.scheme == "http":
+            port = 80
+        elif url_parsed.scheme == "https":
+            port = 443
+        else:
+            raise NotImplementedError()
+
+    hostname = get_domain_normal_form(url_parsed.hostname)
+
+    if url_parsed.scheme in ["http", "https"] and port in Config.COMMON_HTTP_PORTS_FOR_DEDUPLICATION:
+        url_parsed_dict["scheme"] = "http_or_https"
+        url_parsed_dict["netloc"] = hostname + ":0"
+    else:
+        url_parsed_dict["netloc"] = hostname + ":" + str(port)
+
+    return urllib.parse.urlunparse(urllib.parse.ParseResult(**url_parsed_dict))
+
+
+def get_domain_normal_form(domain: str) -> str:
+    """
+    Normalizes a domain - if it starts with www, strips it.
+    """
+    assert ":" not in domain
+
+    if domain.startswith("www."):
+        domain = domain[4:]
+    return domain
```

Comment:
1) why assert?
2) Not to mention this is just a `str.removeprefix` with extra steps and takes 9 LoC.